ngraph/core/reference/include/ngraph/runtime/reference/proposal.hpp

```diff
@@ -0,0 +1,580 @@
+                                      [](const details::ProposalBox<T>& box1,
+                                         const details::ProposalBox<T>& box2) {
+                                          return (box1.score > box2.score);
+                                      });
+                    details::nms(pre_nms_topn,
+                                 is_dead,
+                                 proposals,
+                                 roi_indices,
+                                 num_rois,
+                                 0,
+                                 attrs.nms_thresh,
+                                 attrs.post_nms_topn,
+                                 static_cast<T>(coordinates_offset));
+
+                    details::retrieve_rois(num_rois,
+                                           batch_idx,
+                                           pre_nms_topn,
+                                           proposals,
+                                           roi_indices,
+                                           p_roi_item + batch_idx * attrs.post_nms_topn * 5,
+                                           attrs.post_nms_topn,
+                                           attrs.normalize,
+                                           img_H,
+                                           img_W,
+                                           attrs.clip_after_nms);
+                }
+            }
+
+            template <typename T>
+            void proposal_v4(const T* class_probs,
+                             const T* bbox_deltas,
+                             const T* image_shape,
+                             T* output,
+                             T* out_probs,
+                             const Shape& class_probs_shape,
+                             const Shape& bbox_deltas_shape,
+                             const Shape& image_shape_shape,
+                             const Shape& output_shape,
+                             const Shape& out_probs_shape,
+                             const op::ProposalAttrs& attrs)
+            {
+                const T* p_bottom_item = class_probs;
+                const T* p_d_anchor_item = bbox_deltas;
+                T* p_roi_item = output;
+                T* p_prob_item = attrs.infer_probs ? out_probs : nullptr;
+
+                // bottom shape (batch_size * (2 * num_anchors) * H * W)
+                const unsigned int bottom_H = class_probs_shape[2];
+                const unsigned int bottom_W = class_probs_shape[3];
+                // input image height and width
+                const T img_H = image_shape[0];
+                const T img_W = image_shape[1];
+                // scale factor for H and W, depends on shape of image_shape
+                // can be split into H and W {image_height, image_width, scale_height, scale_width}
+                // or be the same for both   {image_height, image_width, scale_height_and_width}
+                const T scale_H = image_shape[2];
```

Comment:
Looks like the code of this function is duplicated for 90% the code from `proposal_v0` function. Can we think and generalize the approach in order to avoid code duplication?

Reply:
I'll try 👍